File: src/main/java/org/jcodec/containers/mp4/boxes/MetaBox.java
package org.jcodec.containers.mp4.boxes;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class MetaBox extends NodeBox {

    private static final String FOURCC = "meta";

    public MetaBox(Header atom) {
        super(atom);
    }

    public static MetaBox createMetaBox() {
        return new MetaBox(Header.createHeader(fourcc(), 0));
    }

    public Map<String, MetaValue> getKeyedMeta() {
        Map<String, MetaValue> result = new LinkedHashMap<String, MetaValue>();

        IListBox ilst = NodeBox.findFirst(this, IListBox.class, IListBox.fourcc());
        MdtaBox[] keys = NodeBox.findAllPath(this, MdtaBox.class, new String[] { KeysBox.fourcc(), MdtaBox.fourcc() });

        if (ilst == null || keys.length == 0)
            return result;

        for (Entry<Integer, List<Box>> entry : ilst.getValues().entrySet()) {
            Integer index = entry.getKey();
            if (index == null)
                continue;
            DataBox db = getDataBox(entry.getValue());
            if (db == null)
                continue;
            MetaValue value = MetaValue.createOtherWithLocale(db.getType(), db.getLocale(), db.getData());
            if (index > 0 && index <= keys.length) {
                result.put(keys[index - 1].getKey(), value);
            }
        }
        return result;
    }

    private DataBox getDataBox(List<Box> value) {
        for (Box box : value) {
            if (box instanceof DataBox) {
                return (DataBox) box;
            }
        }
        return null;
    }

    public Map<Integer, MetaValue> getItunesMeta() {
        Map<Integer, MetaValue> result = new LinkedHashMap<Integer, MetaValue>();

        IListBox ilst = NodeBox.findFirst(this, IListBox.class, IListBox.fourcc());

        if (ilst == null)
            return result;

        for (Entry<Integer, List<Box>> entry : ilst.getValues().entrySet()) {
            Integer index = entry.getKey();
            if (index == null)
                continue;
            DataBox db = getDataBox(entry.getValue());
            if (db == null)
                continue;
            MetaValue value = MetaValue.createOtherWithLocale(db.getType(), db.getLocale(), db.getData());
            result.put(index, value);
        }
        return result;
    }

    public void setKeyedMeta(Map<String, MetaValue> map) {
        if (map.isEmpty())
            return;
        KeysBox keys = KeysBox.createKeysBox();
        Map<Integer, List<Box>> data = new LinkedHashMap<Integer, List<Box>>();
        int i = 1;
        for (Entry<String, MetaValue> entry : map.entrySet()) {
            keys.add(MdtaBox.createMdtaBox(entry.getKey()));
            MetaValue v = entry.getValue();
            List<Box> children = new ArrayList<Box>();
            children.add(DataBox.createDataBox(v.getType(), v.getLocale(), v.getData()));
            data.put(i, children);
            ++i;
        }
        IListBox ilst = IListBox.createIListBox(data);
        this.replaceBox(keys);
        this.replaceBox(ilst);
    }

    public void setItunesMeta(Map<Integer, MetaValue> map) {
        if (map.isEmpty())
            return;
        Map<Integer, MetaValue> copy = new LinkedHashMap<Integer, MetaValue>();
        copy.putAll(map);
        IListBox ilst = NodeBox.findFirst(this, IListBox.class, IListBox.fourcc());
        Map<Integer, List<Box>> data;
        if (ilst == null) {
            data = new LinkedHashMap<Integer, List<Box>>();
        } else {
            data = ilst.getValues();

            // Updating values
            for (Entry<Integer, List<Box>> entry : data.entrySet()) {
                int index = entry.getKey();
                MetaValue v = copy.get(index);
                if (v != null) {
                    DataBox dataBox = DataBox.createDataBox(v.getType(), v.getLocale(), v.getData());
                    dropChildBox(entry.getValue(), DataBox.fourcc());
                    entry.getValue().add(dataBox);
                    copy.remove(index);
                }
            }
        }

        // Adding values
        for (Entry<Integer, MetaValue> entry : copy.entrySet()) {
            int index = entry.getKey();
            MetaValue v = entry.getValue();
            DataBox dataBox = DataBox.createDataBox(v.getType(), v.getLocale(), v.getData());
            List<Box> children = new ArrayList<Box>();
            data.put(index, children);
            children.add(dataBox);
        }
        
        // Dropping values
        Set<Integer> keySet = new HashSet<Integer>(data.keySet());
        keySet.removeAll(map.keySet());
        for (Integer dropped : keySet) {
            data.remove(dropped);
        }
        
        this.replaceBox(IListBox.createIListBox(data));
    }

    private void dropChildBox(List<Box> children, String fourcc2) {
        ListIterator<Box> listIterator = children.listIterator();
        while (listIterator.hasNext()) {
            Box next = listIterator.next();
            if (fourcc2.equals(next.getFourcc())) {
                listIterator.remove();
            }
        }
    }

    public static String fourcc() {
        return FOURCC;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/MetaBox.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.containers.mp4.boxes.DataBox.Companion.createDataBox
import org.jcodec.containers.mp4.boxes.HandlerBox.Companion.createHandlerBox
import org.jcodec.containers.mp4.boxes.Header.Companion.createHeader
import org.jcodec.containers.mp4.boxes.IListBox.Companion.createIListBox
import org.jcodec.containers.mp4.boxes.KeysBox.Companion.createKeysBox
import org.jcodec.containers.mp4.boxes.MdtaBox.Companion.createMdtaBox
import java.nio.ByteBuffer
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
open class MetaBox(atom: Header) : NodeBox(atom) {
    var keyedMeta: Map<String, MetaValue>
        get() {
            val result: MutableMap<String, MetaValue> = LinkedHashMap()
            val ilst = findFirst(this, IListBox.fourcc()) as IListBox?
            val findAllPath = findAllPath(this, arrayOf(KeysBox.fourcc(), MdtaBox.fourcc()))
            val keys: List<MdtaBox?> = findAllPath.map { it as? MdtaBox? }
            if (ilst == null || keys.size == 0) return result
            for ((key, value1) in ilst.getValues()) {
                val index = key ?: continue
                val db = getDataBox(value1) ?: continue
                val value = MetaValue.createOtherWithLocale(db.type, db.locale, db.data)
                if (index > 0 && index <= keys.size) {
                    result[keys[index - 1]!!.key!!] = value
                }
            }
            return result
        }
        set(map) {
            if (map.isEmpty()) return
            val keys = createKeysBox()
            val data: MutableMap<Int, List<Box>> = LinkedHashMap()
            var i = 1
            for ((key, v) in map) {
                keys.add(createMdtaBox(key))
                val children: MutableList<Box> = ArrayList()
                children.add(createDataBox(v.type, v.locale, v.data))
                data[i] = children
                ++i
            }
            val ilst = createIListBox(data)
            replaceBox(keys)
            replaceBox(ilst)
        }

    private fun getDataBox(value: List<Box>): DataBox? {
        for (box in value) {
            if (box is DataBox) {
                return box
            }
        }
        return null
    }

    override fun parse(input: ByteBuffer) {
        super.parse(input)
    }

    // Updating values

    // Adding values

    // Dropping values
    var itunesMeta: Map<Int, MetaValue>
        get() {
            val result: MutableMap<Int, MetaValue> = LinkedHashMap()
            val ilst = findFirst(this, IListBox.fourcc()) as? IListBox? ?: return result
            for ((key, value1) in ilst.getValues()) {
                val index = key ?: continue
                val db = getDataBox(value1) ?: continue
                val value = MetaValue.createOtherWithLocale(db.type, db.locale, db.data)
                result[index] = value
            }
            return result
        }
        set(map) {
            if (map.isEmpty()) return
            val copy: MutableMap<Int, MetaValue> = LinkedHashMap()
            copy.putAll(map)
            val ilst = findFirst(this, IListBox.fourcc()) as? IListBox?
            val data: MutableMap<Int, MutableList<Box>>
            if (ilst == null) {
                data = LinkedHashMap()
            } else {
                data = ilst.getValues().mapValues { it.value.toMutableList() }.toMutableMap()

                // Updating values
                for ((index, value) in data) {
                    val v = copy[index]
                    if (v != null) {
                        val dataBox = createDataBox(v.type, v.locale, v.data)
                        dropChildBox(value, DataBox.fourcc())
                        value.add(dataBox)
                        copy.remove(index)
                    }
                }
            }

            // Adding values
            for ((index, v) in copy) {
                val dataBox = createDataBox(v.type, v.locale, v.data)
                val children: MutableList<Box> = ArrayList()
                data.put(index, children)
                children.add(dataBox)
            }

            // Dropping values
            val keySet: MutableSet<Int> = HashSet(data.keys)
            keySet.removeAll(map.keys)
            for (dropped in keySet) {
                data.remove(dropped)
            }
            replaceBox(createHandlerBox(null, "mdir", "appl", 0, 0))
            replaceBox(createIListBox(data.mapValues { it.value.toList() }.toMutableMap()))
        }

    private fun dropChildBox(children: MutableList<Box>, fourcc2: String) {
        val listIterator = children.listIterator()
        while (listIterator.hasNext()) {
            val next = listIterator.next()
            if (fourcc2 == next.fourcc) {
                listIterator.remove()
            }
        }
    }

    companion object {
        private const val FOURCC = "meta"
        @JvmStatic
        fun createMetaBox(): MetaBox {
            return MetaBox(createHeader(fourcc(), 0))
        }

        @JvmStatic
        fun fourcc(): String {
            return FOURCC
        }
    }
}